src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -630,19 +632,32 @@ void SumCsrImpl(const nnvm::NodeAttrs& attrs, mshadow::Stream<xpu>* s, const OpC
                 s, num_threads, output->data().dptr<DType>(), in_indptr, in_idx,
                 in_data, sum.dptr_, residual.dptr_, num_rows, num_cols,
                 seg_len);
+            if (normalize) {
+              mshadow::Tensor<xpu, 1, DType> out_data =
+                  output->data().FlatTo1D<xpu, DType>(s);
+              out_data /= scalar<DType>(num_rows);
+            }
           });
         });
       });
     });
   } else if (1 == param.axis[0]) {
     MSHADOW_IDX_TYPE_SWITCH(input.aux_type(kIndPtr), RType, {
-      MSHADOW_TYPE_SWITCH(input.dtype(), DType, {
-        MXNET_ASSIGN_REQ_SWITCH(req, req_type, {
-          const RType* in_indptr = input.aux_data(kIndPtr).dptr<RType>();
-          const DType* in_data = input.data().dptr<DType>();
-          Kernel<SumCsrKernel<req_type, 1>, xpu>::Launch(
-              s, out_data_size, output->data().dptr<DType>(), in_indptr,
-              in_data);
+      MSHADOW_IDX_TYPE_SWITCH(input.aux_type(kIdx), IType, {
```

Comment:
If the mean turns out to be zero, does the op effectively produce a dense result? What does numpy produce in an equivalent dense operation?

Reply:
The operator is mean(csr, axis=0) = dense and thus always produces a dense result.